File: cs160/hangman.py
from random import randint, randrange

def main():
    index = randint(0, 801) # There are 802 pokemon
    pokemon = ''
    with open('pokemon.txt') as fp:
        for i, line in enumerate(fp):
            if i == index:
                pokemon = line.strip()
    gameloop(pokemon)

def gameloop(pokemon):
    progress = ['_' for char in pokemon]
    win = False
    time = 5
    while time > 0 and not win:
        print(''.join(progress))
        guess = input('> ')[0]
        for i, char in enumerate(pokemon):
            if guess == char:
                progress[i] = pokemon[i]
        if guess not in pokemon:
            time -= 1

        win = ''.join(progress) == pokemon

    if win:
        print('That\'s right! It\'s' + pokemon + '.')
        print('Your father and I are so proud of you.')
    else:
        print('You took too long.')
        print('It was ' + pokemon + ' all along.')

if __name__ == '__main__':
    main()

File: cs160/hangman.exs
defmodule Hangman do
  def main() do
    pokemon = File.read!("pokemon.txt")
      |> String.split()
      |> Enum.random()
    pokemon = to_charlist(pokemon)
    progress = Enum.map(pokemon, fn x -> "_" end)
   
    gameloop(pokemon, progress, 5)
  end

  def gameloop(pokemon, progress, time) do
    IO.puts progress
    guess = hd(to_charlist(IO.gets "> "))
    matches = Enum.map(pokemon, fn x -> if x == guess, do: x, else: "_" end)
    new_progress = Enum.zip(matches, progress)
      |> Enum.map(fn {x, y} -> Enum.min([x,y]) end)
    cond do
      new_progress == pokemon -> win pokemon
      time <= 0 -> lose pokemon
      Enum.find(pokemon, fn x -> x == guess end) ->
        gameloop(pokemon, new_progress, time)
      true ->
        gameloop(pokemon, new_progress, time - 1)
    end
  end

  def win(pokemon) do
    IO.puts("That's right! It's " <> to_string(pokemon) <> ".")
    IO.puts("Your father and I are so proud of you.")
  end

  def lose(pokemon) do
    IO.puts("You took too long.")
    IO.puts("It was " <> to_string(pokemon) <> " all along.")
  end
end

Hangman.main()
